File: src/services/danmaku/messagePreview.ts
import { chatFileKind, formatFileSize, parseChatAttachment } from '../chat/fileAttachment';
import { sniffImageMime } from '../chat/imageData';

export type PreviewKind = 'text' | 'image' | 'video' | 'voice' | 'audio' | 'file';
export interface PreviewMessage { type?: string; content: string; imageData?: string; attachment?: unknown; recalled?: boolean }
export interface MessagePreview { kind: PreviewKind; text: string; detail?: string; image?: string }
export function messagePreview(message: PreviewMessage): MessagePreview {
  if (message.recalled) return { kind: 'text', text: '[消息已撤回]' };
  if (!message.type || message.type === 'text') return { kind: 'text', text: message.content.replace(/^> \[reply:[^\]]+\]\s*/, '> ') };
  if (message.type === 'image') return { kind: 'image', text: '[图片 / 表情]', image: message.imageData };
  if (message.type === 'voice') {
    let duration = 0;
    try { duration = Number(JSON.parse(message.content).duration); } catch { /* legacy voice */ }
    return { kind: 'voice', text: '语音消息', detail: Number.isFinite(duration) && duration > 0 && duration <= 120 ? `${Math.ceil(duration)} 秒` : undefined };
  }
  if (message.type === 'file') {
    const file = parseChatAttachment(message.attachment ?? message.content);
    if (!file) return { kind: 'file', text: '[附件不可用]' };
    const kind = chatFileKind(file);
    return { kind: kind === 'image' || kind === 'video' || kind === 'audio' ? kind : 'file', text: file.name,
      detail: `${file.name.split('.').pop()?.toUpperCase() ?? 'FILE'} · ${formatFileSize(file.size)}` };
  }
  return { kind: 'file', text: '[不支持的消息]' };
}

/** A bounded thumbnail crosses the overlay event boundary, never attachment metadata or remote URLs. */
export async function visualThumbnail(src: string, kind: 'image' | 'video'): Promise<string> {
  if (kind === 'image') {
    const response = await fetch(src, { signal: AbortSignal.timeout(12000) });
    if (!response.ok) throw new Error('Preview unavailable');
    const blob = await response.blob();
    const mime = sniffImageMime(new Uint8Array(await blob.slice(0, 32).arrayBuffer()));
    if (blob.size <= 2 * 1024 * 1024 && mime) {
      return await new Promise((resolve, reject) => {
        const reader = new FileReader(); reader.onload = () => resolve(String(reader.result)); reader.onerror = reject; reader.readAsDataURL(blob.slice(0, blob.size, mime));
      });
    }
  }
  return await new Promise((resolve, reject) => {
    const media = kind === 'video' ? document.createElement('video') : new Image();
    const timer = window.setTimeout(() => finish(new Error('Preview timeout')), 12000);
    const finish = (error?: Error, value?: string) => {
      clearTimeout(timer); media.onload = null; media.onerror = null;
      if (media instanceof HTMLVideoElement) { media.onloadeddata = null; media.onseeked = null; media.removeAttribute('src'); media.load(); }
      if (error) reject(error); else resolve(value!);
    };
    const render = () => {
      try {
        const w = media instanceof HTMLVideoElement ? media.videoWidth : media.naturalWidth;
        const h = media instanceof HTMLVideoElement ? media.videoHeight : media.naturalHeight;
        if (!w || !h) throw new Error('No visual track');
        const scale = Math.min(1, 320 / w, 180 / h);
        const canvas = document.createElement('canvas'); canvas.width = Math.max(1, Math.round(w * scale)); canvas.height = Math.max(1, Math.round(h * scale));
        const context = canvas.getContext('2d'); if (!context) throw new Error('No canvas');
        context.drawImage(media, 0, 0, canvas.width, canvas.height); finish(undefined, canvas.toDataURL('image/jpeg', .82));
      } catch (error) { finish(error as Error); }
    };
    media.onerror = () => finish(new Error('Preview unavailable'));
    if (media instanceof HTMLVideoElement) { media.muted = true; media.preload = 'auto'; media.onloadeddata = render; }
    else media.onload = render;
    media.src = src;
  });
}

File: MCTier-Android/app/src/main/java/top/pmh13/mctier/data/MessagePreview.kt
package top.pmh13.mctier.data

import org.json.JSONObject

data class MessagePreview(val kind: String, val text: String, val detail: String = "")

fun messagePreview(message: ChatMessage): MessagePreview {
    if (message.recalled) return MessagePreview("text", "[消息已撤回]")
    return when (message.type) {
        "text" -> MessagePreview("text", message.content.replaceFirst(Regex("^> \\[reply:[^]]+]\\s*"), "> "))
        "image" -> MessagePreview("image", "[图片 / 表情]")
        "voice" -> {
            val duration = runCatching { JSONObject(message.content).optDouble("duration", 0.0) }.getOrDefault(0.0)
            MessagePreview("voice", "语音消息", if (duration.isFinite() && duration > 0 && duration <= 120) "${kotlin.math.ceil(duration).toInt()} 秒" else "")
        }
        "file" -> {
            val file = message.attachment?.takeIf(::validChatAttachment) ?: return MessagePreview("file", "[附件不可用]")
            val kind = chatAttachmentKind(file).takeIf { it in setOf("image", "video", "audio") } ?: "file"
            val size = when { file.size < 1024 -> "${file.size} B"; file.size < 1048576 -> "${file.size / 1024} KB"; else -> "${file.size / 1048576} MB" }
            MessagePreview(kind, file.name, "${file.name.substringAfterLast('.', "FILE").uppercase()} · $size")
        }
        else -> MessagePreview("file", "[不支持的消息]")
    }
}
